### ISSUE TITLE
您的项目和parlai项目

### ISSUE THREAD

**Issue by chrislouis0106** [THUDM/KBRD#13]
我阅读您的代码，也看了parlai源码。但是我不太知道你是如何在parlai源码的基础上代码你的任务，所以我想问你是如何在parlai源码的基础上插入式开发自己的任务，或者在parlai源码的基础上你的任务框架是什么样的，（论文上的模型框架我是可以理解的）。

**Comment by qibinc** [comment]
Hi @Ashappyboy ,

感谢对此研究抱有兴趣。

本 repo 中把 parlai 精简成如下三部分，相信你已经注意到：

```
core/
agents/
tasks/
```

1. `core/` 中主要是各种 task、model 都共享的代码，比如 dataloader, training loss, eval metrics 等等。这里面除了一些必要的 hack，我们没有其他改动，可以跳过；
2. `tasks/` 是各类任务的集合。原 ParlAI 不仅有对话，也有 QA 之类，都是在这里添加实现的。我们按照parlai的 task 的写法，在里面加了 parlai 数据集，对 redial 数据集的读取、处理等都可以在 parlai/tasks/redial/agents.py 里面找到。除此之外，由于本文中用到的 knowledge graph 是一个跟 redial 有关的 dbpedia 子图，这部分代码也可以在 parlai/tasks/redial/ 下找到。
3. `agents/` 里面是我们以及 baseline 的模型。autorec/ 是 recommender baseline，transformer/ 是 dialog baseline，kbrd/ 是 our recommender，tranformer_rec/ 是 our dialog，最终的对话推荐由 transformer_rec/ 实现，你可以看到代码中用到了 kbrd/ 里的推荐模型。由于 recommender 事实上只依赖于对话 contents，而不是对话模型，因此在实际中我们采用了将两者的 training 分离（即先训好 recommender），可以加快之后 dialog model 的收敛。

Best wishes,
Qibin

**Comment by chrislouis0106** [comment]
谢谢您的耐心回复，你这个回复真的让我更加理解了您的工作，同时在一个issue中，您解释了这个测试和验证过程，我已近基本上可以理解整个模型和流程了。但是我有些一些小问题想请教您：
1、在您的scripts文件夹下的score.py文件，里面的参数我不太明白，可以给一个调用案例或者说明一个用途吗
2、在task下，很明显您的gen_rippleset.py和extract_entities.py是您用来测试的嘛？因为我看到好像这两个文件在整个redial任务中是没有用处的
Thank you very much

**Comment by qibinc** [comment]
Hi @Ashappyboy ,

1. `score.py` 是用来算多次实验的 mean/std，以及比较两种 model 结果差异的显著性。比如我们跑了 32 次 baseline（`bash scripts/baseline.sh 32 0`） 和 ours (`bash scripts/both.sh 32 1`)。等结果出来之后，就可以调用 `python scripts/score.py --name-1 saved/both_rgcn --name-2 saved/release_baseline --num 32 --metric recall@50`，结果如下

```bash
[0.3292, 0.3447, 0.3331, 0.3463, 0.3325, 0.3419, 0.3401, 0.3413, 0.3393, 0.3395, 0.3449, 0.3345, 0.328, 0.3335, 0.3246, 0.3329, 0.3385, 0.3528, 0.3335, 0.3399, 0.3417, 0.3395, 0.3335, 0.328, 0.3244, 0.3508, 0.3429, 0.3329, 0.3413, 0.3423, 0.3433]
0.3377935483870968
0.006955894946426603
[0.2514, 0.2608, 0.26, 0.2584, 0.2536, 0.2586, 0.2499, 0.2508, 0.2606, 0.2493, 0.26, 0.263, 0.2648, 0.263, 0.2399, 0.2463, 0.2532, 0.2512, 0.2626, 0.2602, 0.2594, 0.2552, 0.2546, 0.2534, 0.2546, 0.2564, 0.254, 0.259, 0.2622, 0.2608, 0.2477]
0.25596451612903226
0.005713512471839794
Ttest_indResult(statistic=49.79074319303046, pvalue=1.6600794094932858e-50)
```

2. 是的，这些文件用来做了一些预处理，生成的东西应该保存在 `text_dict.pkl` 里，因为直接下载下来了就不用用到。保留着是为了如果跑别的的数据集需要重新处理之类。

**Comment by chrislouis0106** [comment]
谢谢您我明白啦；我关闭这个对话啦，如果有新的问题，我在重新发起一个。
